### PR TITLE
opae: add metrics threshold api

### PR DESCRIPTION
**Pull request by anandaravuri**
metrics threshold api enum threshold info and values

### CODE REVIEW

File: libopae/plugins/xfpga/metrics/metrics_int.h [hunk 1]
```diff
@@ -237,4 +237,32 @@ fpga_result get_metric_data_info(const char *group_name,
 		uint64_t size,
 		fpga_metric_metadata *metric_data);
 
+
+fpga_result xfpga_bmcLoadSDRs(struct _fpga_handle *_handle,
+	bmc_sdr_handle *records,
+	uint32_t *num_sensors);
+
+	fpga_result xfpga_bmcDestroySDRs(struct _fpga_handle *_handle,
```
Comment (jelonanX) [inline libopae/plugins/xfpga/metrics/metrics_int.h]: Weird Indentation.

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+
+
+			// Lower Non-Recoverable Threshold
+			if (details.thresholds.lower_nr_thresh.is_valid) {
+
+				e = strncpy_s(metric_thresholds[x].lower_nr_threshold.threshold_name,
+					sizeof(metric_thresholds[x].lower_nr_threshold.threshold_name),
+					LOWER_NR_THRESHOLD, SYSFS_PATH_MAX);
+				if (EOK != e) {
+					FPGA_ERR("Failed to copy threshold name");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: shouldn't this affect the return code?

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+
+
+			// Lower Critical Threshold
+			if (details.thresholds.lower_c_thresh.is_valid) {
+
+				e = strncpy_s(metric_thresholds[x].lower_c_threshold.threshold_name,
+					sizeof(metric_thresholds[x].lower_c_threshold.threshold_name),
+					LOWER_C_THRESHOLD, SYSFS_PATH_MAX);
+				if (EOK != e) {
+					FPGA_ERR("Failed to copy threshold name");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: shouldn't this affect the return code?

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+			}
+
+			// Lower Non-Critical Threshold
+			if (details.thresholds.lower_nc_thresh.is_valid) {
+
+				e = strncpy_s(metric_thresholds[x].lower_nc_threshold.threshold_name,
+					sizeof(metric_thresholds[x].lower_nc_threshold.threshold_name),
+					LOWER_NC_THRESHOLD, SYSFS_PATH_MAX);
+				if (EOK != e) {
+					FPGA_ERR("Failed to copy threshold name");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: shouldn't this affect the return code?

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+				metric_thresholds[x].lower_nc_threshold.value = details.thresholds.lower_nc_thresh.value;
+				metric_thresholds[x].lower_nc_threshold.is_valid = true;
+			}
+
+		} // for loop end
+
+	} // endif
+
+destroy_values:
+	result = xfpga_bmcDestroySensorValues(_handle, &values);
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: Any result from above is overwritten here.

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+	if (result != FPGA_OK) {
+		FPGA_MSG("Failed to Destroy Sensor value.");
+	}
+
+destroy_sdr:
+	result = xfpga_bmcDestroySDRs(_handle, &records);
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: Any result from above is overwritten here.

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+
+	if (handle == NULL ||
+		num_thresholds == NULL) {
+		FPGA_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+
+	struct _fpga_handle *_handle = (struct _fpga_handle *)handle;
+
+	if (_handle->bmc_handle == NULL)
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: Need to lock the _handle around this check and the subsequent assignment of _handle->bmc_handle.

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+	}
+
+	// read thresholds
+	for (i = 0; i < pglob.gl_pathc; i++) {
+
+		// Sensor name
+		result = read_sensor_sysfs_file(pglob.gl_pathv[i], SENSOR_SYSFS_NAME, (void **)&tmp, &tot_bytes);
+		if (FPGA_OK != result) {
+			if (tmp) {
+				free(tmp);
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: Set tmp to NULL to prevent double-free?

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+			}
+			continue;
+		}
+
+		memset_s(&metric_thresholds[i].metric_name, sizeof(metric_thresholds[i].metric_name), 0);
+		e = strncpy_s(metric_thresholds[i].metric_name, sizeof(metric_thresholds[i].metric_name),
+			(char *)tmp, strnlen_s((char *)tmp, SYSFS_PATH_MAX));
+		if (EOK != e) {
+			FPGA_ERR("Failed to copy threshold name");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: Shouldn't this affect the return code?

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+		}
+		metric_thresholds[i].metric_name[strlen(metric_thresholds[i].metric_name) - 1] = '\0';
+
+
+		// Upper Critical Threshold
+		e = strncpy_s(metric_thresholds[i].upper_c_threshold.threshold_name,
+			sizeof(metric_thresholds[i].upper_c_threshold.threshold_name),
+			UPPER_C_THRESHOLD, SYSFS_PATH_MAX);
+		if (EOK != e) {
+			FPGA_ERR("Failed to copy threshold name");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: Shouldn't this affect the return code?

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+			metric_thresholds[i].upper_c_threshold.value = ((double)value / MILLI);
+			metric_thresholds[i].upper_c_threshold.is_valid = true;
+		}
+
+		// Upper Non-Critical Threshold
+		e = strncpy_s(metric_thresholds[i].upper_nc_threshold.threshold_name,
+			sizeof(metric_thresholds[i].upper_nc_threshold.threshold_name),
+			UPPER_NC_THRESHOLD, SYSFS_PATH_MAX);
+		if (EOK != e) {
+			FPGA_ERR("Failed to copy threshold name");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: Shouldn't this affect the return code?

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+			metric_thresholds[i].upper_nc_threshold.is_valid = true;
+		}
+
+
+		// Lower Critical Threshold
+		e = strncpy_s(metric_thresholds[i].upper_nc_threshold.threshold_name,
+			sizeof(metric_thresholds[i].lower_c_threshold.threshold_name),
+			LOWER_C_THRESHOLD, SYSFS_PATH_MAX);
+		if (EOK != e) {
+			FPGA_ERR("Failed to copy threshold name");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: Shouldn't this affect the return code?

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+			metric_thresholds[i].lower_c_threshold.value = ((double)value / MILLI);
+			metric_thresholds[i].lower_c_threshold.is_valid = true;
+		}
+
+		// Lower Non-Critical Threshold
+		e = strncpy_s(metric_thresholds[i].lower_nc_threshold.threshold_name,
+			sizeof(metric_thresholds[i].lower_nc_threshold.threshold_name),
+			LOWER_NC_THRESHOLD, SYSFS_PATH_MAX);
+		if (EOK != e) {
+			FPGA_ERR("Failed to copy threshold name");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: Shouldn't this affect the return code?

File: libopae/plugins/xfpga/metrics/threshold.c [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+			metric_thresholds[i].lower_nc_threshold.value = ((double)value / MILLI);
+			metric_thresholds[i].lower_nc_threshold.is_valid = true;
+		}
+
+		// Lower Non-Critical Threshold
+		e = strncpy_s(metric_thresholds[i].hysteresis.threshold_name,
+			sizeof(metric_thresholds[i].hysteresis.threshold_name),
+			SYSFS_HYSTERESIS, SYSFS_PATH_MAX);
+		if (EOK != e) {
+			FPGA_ERR("Failed to copy threshold name");
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.c]: Shouldn't this affect the return code?

File: libopae/plugins/xfpga/metrics/threshold.h [hunk 15]
```diff
@@ -0,0 +1,63 @@
+// Copyright(c) 2018, Intel Corporation
```
Comment (tswhison) [inline libopae/plugins/xfpga/metrics/threshold.h]: s/2018/2019/

File: libopae/plugins/xfpga/metrics/threshold.c (hunk 7 above)
Comment (tswhison) [reply libopae/plugins/xfpga/metrics/threshold.c]: ie
```c
lock(&_handle->lock);
if (_handle->bmc_handle == NULL)
   _handle->bmc_handle = ...
unlock(&_handle->lock);
```

**Pull request merged**